File: CS50X/pset6/credit.py
# Name: Credit.py
# Author: Robin Goyal
# Last-Modified: May 24, 2017
# Purpose: Determine if a credit card is valid 
#          the type of credit card

import cs50

def main():
    
    # Get credit card number and split string into integer digits in array
    print("Credit Card Number: ", end = "")
    cardInput = str(cs50.get_int())
    cardNumber = [int(i) for i in cardInput]
    
    # result from luhns algorithm
    total = luhn(cardNumber)
        
    # Valid credit cards final digit is 0
    if (total % 10 == 0):
        
        # Visa condition
        if cardInput[0] == "4":
            print("VISA")
            
        # Amex condition
        elif cardInput[0:2] == "34" or cardInput[0:2] == "37":
            print("AMEX")
        
        # Mastercard condition
        elif cardInput[0:2] == "51" or cardInput[0:2] == "52" or cardInput[0:2] == "53" or \
            cardInput[0:2] == "54" or cardInput[0:2] == "55":
            print("MASTERCARD")
    else:
        print("INVALID")
    
    
# Luhns algorithm
def luhn(card):
    total = 0
    
    for i in range(len(card) - 2, -1, -2):
        temp = card[i] * 2
        if temp > 9:
            temp -= 9
        total += temp
        del card[i]
        
    total += sum(card)
    return total

if __name__ == "__main__":
    main()

File: CS50XR2/pset1/credit/credit.c
/* Name: credit.c
   Author: Robin Goyal
   Last-Modified: January 12, 2018
   Purpose: Determine whether a credit card number is valid
*/

#include <stdio.h>
#include <cs50.h>
#include <math.h>

int main(void)
{

    // Get credit card number
    long long card_number = get_long_long("Number: ");

    // Calculate number of digits in credit card
    int card_len = floor(log10(card_number)) + 1;


    // Declare array of size card_len and convert card digits to array
    int card_digits[card_len];

    for (int i = 0; i < card_len; i++)
    {
        card_digits[card_len - i - 1] = card_number % 10;
        card_number /= 10;
    }

    // Length of card number is even
    bool card_len_even = card_len % 2 == 0;

    // Calculate the sum specificed by luhn's algorithm
    int luhn_sum = 0;

    // Iterate over the individual digits of the card
    for (int i = 0; i < card_len; i++)
    {
        if ((card_len_even && i % 2 == 0) || (!(card_len_even) && i % 2 == 1))
        {

            // Add individual digits of the temporary number to luhn sum
            int temp = 2 * card_digits[i];

            while (temp > 0)
            {
                luhn_sum += temp % 10;
                temp /= 10;
            }
        }

        else
        {
            luhn_sum += card_digits[i];
        }
    }


    // Check if credit card number is valid and print the type of card
    if (luhn_sum % 10 == 0)
    {
        if (card_digits[0] == 3 && (card_digits[1] == 4 || card_digits[1] == 7))
        {
            printf("AMEX\n");
        }

        else if (card_digits[0] == 4)
        {
            printf("VISA\n");
        }

        else
        {
            printf("MASTERCARD\n");
        }
    }

    else
    {
        printf("INVALID\n");
    }
}
